File: Java/src/_0741_Cherry_Pickup/Solution.java
package _0741_Cherry_Pickup;

import java.util.Arrays;

/**
 * Date: 2021/6/29 22:17
 * 执行用时：43 ms, 在所有 Java 提交中击败了13.41%的用户
 * 内存消耗：46.1 MB, 在所有 Java 提交中击败了10.36%的用户
 */
public class Solution {

    static final int N = 55;
    int[][][] f = new int[N][N][N * 2];

    public int cherryPickup(int[][] grid) {
        int n = grid.length;
        for (int i = 0; i < N; i++)
            for (int j = 0; j < N; j++)
                Arrays.fill(f[i][j], (int) -1e8);
        f[1][1][2] = grid[0][0];
        for (int k = 3; k <= n * 2; k++)
            for (int i = Math.max(1, k - n); i <= Math.min(n, k - 1); i++)
                for (int j = Math.max(1, k - n); j <= Math.min(n, k - 1); j++) {
                    if (grid[i - 1][k - i - 1] == -1 || grid[j - 1][k - j - 1] == -1) continue;
                    int t = grid[i - 1][k - i - 1];
                    if (i != j) t += grid[j - 1][k - j - 1];
                    for (int a = i - 1; a <= i; a++)
                        for (int b = j - 1; b <= j; b++)
                            f[i][j][k] = Math.max(f[i][j][k], f[a][b][k - 1] + t);
                }
        return Math.max(0, f[n][n][n * 2]);
    }

    public static void main(String[] args) {

        int[][] grid = {
                {0, 1, -1},
                {1, 0, -1},
                {1, 1, 1},
        };
        System.out.println((new Solution()).cherryPickup(grid));  // 5
    }
}

File: Python/_0741_Cherry_Pickup/Solution.py
# coding=utf-8
# Date: 2021/6/29 22:24
from typing import List


# TLE
class Solution:
    def cherryPickup(self, grid: List[List[int]]) -> int:
        n = len(grid)
        N = 55
        f = [[[(int)(-1e8) for _ in range(N * 2)] for _ in range(N)] for _ in range(N)]
        f[1][1][2] = grid[0][0]
        for k in range(3, n * 2 + 1):
            for i in range(max(1, k - n), min(n, k - 1) + 1):
                for j in range(max(1, k - n), min(n, k - 1) + 1):
                    if grid[i - 1][k - i - 1] == -1 or grid[j - 1][k - j - 1] == -1:
                        continue
                    t = grid[i - 1][k - i - 1]
                    if i != j:
                        t += grid[j - 1][k - j - 1]
                    for a in range(i - 1, i + 1):
                        for b in range(j - 1, j + 1):
                            f[i][j][k] = max(f[i][j][k], f[a][b][k - 1] + t)
        return max(0, f[n][n][n * 2])


if __name__ == "__main__":
    print(Solution().cherryPickup([[0, 1, -1], [1, 0, -1], [1, 1, 1]]))  # 5
